crates/dekaf/src/session.rs

```diff
@@ -713,6 +738,53 @@ impl Session {
                     tracing::debug!(collection = ?&key.0, "Collection doesn't exist!");
                     continue; // Collection doesn't exist.
                 };
+
+                // Validate consumer's leader epoch against current collection epoch
+                if partition_request.current_leader_epoch >= 0 {
+                    if partition_request.current_leader_epoch
+                        < collection.binding_backfill_counter as i32
+                    {
+                        metrics::counter!(
+                            "dekaf_fetch_requests",
+                            "topic_name" => key.0.to_string(),
+                            "partition_index" => key.1.to_string(),
+                            "task_name" => task_name.to_string(),
+                            "state" => "fenced_leader_epoch"
+                        )
+                        .increment(1);
+                        tracing::info!(
+                            collection = ?&key.0,
+                            partition = partition_request.partition,
+                            consumer_epoch = partition_request.current_leader_epoch,
+                            current_epoch = collection.binding_backfill_counter,
+                            "Consumer epoch is stale, skipping read start"
+                        );
+                        // Remove stale pending read if it exists. Error will be returned during poll phase
```

Comment:
nit: this action-at-a-distance is confusing / hard to follow. Is it possible to fail faster?